crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,193 @@
+            trace!("process exited: {:?}", out);
+        }
+
+        ProcessManager { state: Closed }
+    }
+
+    /// Stop the process manager, waiting for all child processes to exit.
+    ///
+    /// If you want to set a timeout, use `tokio::time::timeout` and
+    /// `Self::stop` if the timeout elapses.
+    pub async fn wait(self) -> ProcessManager<Closed> {
+        let mut set = JoinSet::new();
+
+        for mut child in self.state.0.into_iter() {
+            set.spawn(async move { child.wait().await });
+        }
+
+        debug!("waiting for {} processes to exit", set.len());
+
+        while let Some(out) = set.join_next().await {
+            trace!("process exited: {:?}", out);
+        }
+
+        ProcessManager { state: Closed }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use tokio::{process::Command, time::sleep};
+    use tracing_test::traced_test;
+
+    use super::*;
+
+    fn get_command() -> Command {
+        let mut cmd = Command::new("node");
+        cmd.arg("./test/scripts/sleep_5_interruptable.js");
+        cmd
+    }
+
+    #[tokio::test]
+    async fn test_basic() {
+        let mut manager = ProcessManager::new().start();
+        manager.spawn(get_command(), Duration::from_secs(2));
+        manager.stop().await;
+    }
+
+    #[tokio::test]
+    async fn test_multiple() {
+        let mut manager = ProcessManager::new().start();
+
+        manager.spawn(get_command(), Duration::from_secs(2));
+        manager.spawn(get_command(), Duration::from_secs(2));
+        manager.spawn(get_command(), Duration::from_secs(2));
+
+        sleep(Duration::from_millis(100)).await;
+
+        manager.stop().await;
+    }
+
+    #[tokio::test]
+    async fn test_closed() {
+        let mut manager = ProcessManager::new().start();
+        manager.spawn(get_command(), Duration::from_secs(2));
+
+        let mut manager = manager.stop().await.start();
+        manager.spawn(get_command(), Duration::from_secs(2));
+
+        sleep(Duration::from_millis(100)).await;
+
+        manager.stop().await;
+    }
+
+    #[tokio::test]
+    async fn test_exit_code() {
+        let mut manager = ProcessManager::new().start();
+        let mut child = manager.spawn(get_command(), Duration::from_secs(2));
+
+        sleep(Duration::from_millis(100)).await;
+
+        let code = child.wait().await;
+        assert_eq!(code, Some(0));
+
+        manager.stop().await;
+    }
+
+    #[tokio::test]
+    #[traced_test]
+    async fn test_message_after_stop() {
+        let mut manager = ProcessManager::new().start();
+        let mut child = manager.spawn(get_command(), Duration::from_secs(2));
+
+        sleep(Duration::from_millis(100)).await;
+
+        let code = child.wait().await;
+        assert_eq!(code, Some(0));
+
+        manager.stop().await;
+
+        // this is idempotent, so calling it after the manager is stopped is ok
+        child.kill().await;
+
+        let code = child.wait().await;
+        assert_eq!(code, None);
```

Comment:
I think would be useful to keep the behavior we had in Go and have a sentinel error for when the child was killed by us vs. killed by someone else (or at least we had the intent on killing it)

Reply:
Added